### PR TITLE
Fix initialization order in KafkaClient

### LINKED ISSUE #2118: Weird error after bad producer initialization
Hi, 
I'm getting this weird error after the GC tries to clean the memory and trigger the `__del__` method of `KafkaClient`. When the producer is not successfully initialized,  the property `_closed` is not ready in the `KafkaClient` object

this snippet if to illustrate the point:
```python
from kafka import KafkaProducer
import os
import gc

def test_bad_kafka_initialization():
    connect_str = os.getenv('BROKER_URL')
    try:
        producer = KafkaProducer(bootstrap_servers=connect_str)
    except Exception as e:
        gc.collect()
        assert isinstance(e, TypeError)
        raise e
```
if the `BROKER_URL` env variable is not defined is pretty normal raise a `TypeError` exception but after that, another ignored exception shows up :

```
Exception ignored in: <function KafkaClient.__del__ at 0x10ecc04c0>
Traceback (most recent call last):
  File "..../python3.8/site-packages/kafka/client_async.py", line 443, in __del__
  File "..../python3.8/site-packages/kafka/client_async.py", line 417, in _close
AttributeError: 'KafkaClient' object has no attribute '_closed'
```
this happens because in `KafkaClient.__init__`  are properties defined after too after all initialization:
```        
 self.cluster = ClusterMetadata(**self.config)   -> here the initialization broke
 .....
 self._closed = False
 self._wake_r, self._wake_w = socket.socketpair()
 self._selector = self.config['selector']()

### PR DESCRIPTION
**Pull request by pecalleja**
this resolve #2118

---
This change is [Reviewable](https://reviewable.io/reviews/dpkp/kafka-python/2119)

### CODE REVIEW

File: kafka/client_async.py [hunk 1]
```diff
@@ -200,19 +200,21 @@ def __init__(self, **configs):
         for key in self.config:
             if key in configs:
                 self.config[key] = configs[key]
-
+        self._closed = False
+        self._wake_r, self._wake_w = socket.socketpair()
+        self._selector = self.config['selector']()
```
Comment (jeffwidman) [inline kafka/client_async.py:205]: It'd be good to have a comment here noting that the above properties should be top so as to not break when `__del__` is called.

**Review by jeffwidman — APPROVED** [review]
LGTM, although could you add a comment? It'd be all too easy for this problem to creep back in later from someone who doesn't realize the order matters for these specific attributes...

File: kafka/client_async.py (hunk 1 above)
Comment (pecalleja) [reply kafka/client_async.py:208]: sure !!!  done.

File: kafka/client_async.py [hunk 2]
```diff
@@ -201,18 +201,24 @@ def __init__(self, **configs):
...
+        # because they are used when __del__ method is called
+        self._closed = False
+        self._wake_r, self._wake_w = socket.socketpair()
+        self._selector = self.config['selector']()
+
         self.cluster = ClusterMetadata(**self.config)
         self._topics = set()  # empty set will fetch all topic metadata
         self._metadata_refresh_in_progress = False
-        self._selector = self.config['selector']()
+
```
Comment (jeffwidman) [inline kafka/client_async.py:213]: Wouldn't you want to delete these new lines?

File: kafka/client_async.py [hunk 3]
```diff
@@ -226,7 +232,7 @@ def __init__(self, **configs):
 
         self._selector.register(self._wake_r, selectors.EVENT_READ)
         self._idle_expiry_manager = IdleConnectionManager(self.config['connections_max_idle_ms'])
-        self._closed = False
+
```
Comment (jeffwidman) [inline kafka/client_async.py:235]: Wouldn't you want to delete these new lines?

File: kafka/client_async.py [hunk 4]
```diff
@@ -201,18 +201,24 @@ def __init__(self, **configs):
...
         self._conns = Dict()  # object to support weakrefs
         self._api_versions = None
         self._connecting = set()
         self._sending = set()
         self._refresh_on_disconnects = True
         self._last_bootstrap = 0
         self._bootstrap_fails = 0
-        self._wake_r, self._wake_w = socket.socketpair()
+
```
Comment (jeffwidman) [inline kafka/client_async.py:221]: Wouldn't you want to delete these new lines?

File: kafka/client_async.py [hunk 5]
```diff
@@ -201,18 +201,24 @@ def __init__(self, **configs):
             if key in configs:
                 self.config[key] = configs[key]
 
+        # these properties need to be set on top of the initialization pipeline
+        # because they are used when __del__ method is called
+        self._closed = False
+        self._wake_r, self._wake_w = socket.socketpair()
+        self._selector = self.config['selector']()
+
```
Comment (jeffwidman) [inline kafka/client_async.py:209]: Keep this newline as part of keeping the "required initialization section" separate

File: kafka/client_async.py (hunk 2 above)
Comment (pecalleja) [reply kafka/client_async.py]: 👍

File: kafka/client_async.py (hunk 3 above)
Comment (pecalleja) [reply kafka/client_async.py]: 👍

File: kafka/client_async.py (hunk 4 above)
Comment (pecalleja) [reply kafka/client_async.py]: 👍

File: kafka/client_async.py (hunk 5 above)
Comment (pecalleja) [reply kafka/client_async.py:209]: 👍

**Review by jeffwidman — APPROVED** [review]

**Pull request merged**